engine/src/flutter/display_list/dl_canvas.cc

```diff
@@ -4,22 +4,132 @@
 
 #include "flutter/display_list/dl_canvas.h"
 
-#include "flutter/display_list/geometry/dl_geometry_conversions.h"
-#include "flutter/third_party/skia/include/core/SkPoint3.h"
-#include "flutter/third_party/skia/include/utils/SkShadowUtils.h"
+namespace {
+
+// ShadowBounds code adapted from SkShadowUtils using the Directional flag.
+
+using DlScalar = flutter::DlScalar;
+using DlVector3 = flutter::DlVector3;
+using DlVector2 = flutter::DlVector2;
+using DlRect = flutter::DlRect;
+using DlMatrix = flutter::DlMatrix;
+
+static constexpr DlScalar kAmbientHeightFactor = 1.0f / 128.0f;
+static constexpr DlScalar kAmbientGeomFactor = 64.0f;
+// Assuming that we have a light height of 600 for the spot shadow, the spot
+// values will reach their maximum at a height of approximately 292.3077.
+// We'll round up to 300 to keep it simple.
+static constexpr DlScalar kMaxAmbientRadius =
+    300.0f * kAmbientHeightFactor * kAmbientGeomFactor;
+
+inline DlScalar AmbientBlurRadius(DlScalar height) {
+  return std::min(height * kAmbientHeightFactor * kAmbientGeomFactor,
+                  kMaxAmbientRadius);
+}
+
+struct DrawShadowRec {
+  DlVector3 light_position;
+  DlScalar light_radius;
+  DlScalar occluder_z;
+};
+
+static inline float DivideAndPin(float numer,
+                                 float denom,
+                                 float min,
+                                 float max) {
+  float result = std::clamp(numer / denom, min, max);
+  // ensure that clamp handled non-finites correctly
+  FML_DCHECK(result >= min && result <= max);
```

Comment:
Check seems redundant?

Reply:
Skia had it. I think it can run into trouble with non-finites...?